stix2/registration.py

```diff
@@ -3,10 +3,83 @@
 from . import registry
 from .base import _DomainObject, _Observable
 from .exceptions import DuplicateRegistrationError
-from .properties import _validate_type
+from .properties import (
+    _validate_type, ListProperty, ObjectReferenceProperty, ReferenceProperty,
+)
 from .utils import PREFIX_21_REGEX, get_class_hierarchy_names
 from .version import DEFAULT_VERSION
 
+# Properties ending in "_ref/s" need to be instances of specific types to meet
+# our interpretation of section 3.1 of the spec
+_VERSION_REF_TYPES_MAP = {
+    # TODO: The 2.0 code seems to use `ReferenceProperty`s - is that okay?
+    "2.0": (ObjectReferenceProperty, ReferenceProperty, ),
+    "2.1": (ReferenceProperty, ),
+}
+# Any unknown versions are presumed to be newer than STIX 2.1 so we'll define
+# this to enforce the 2.1 compatible type in such situations
+_UNKNOWN_VERSION_REF_TYPES = _VERSION_REF_TYPES_MAP["2.1"]
+
+
+def _validate_ref_props(props_map, version):
+    """
+    Validate that reference properties contain an expected type.
+
+    Args:
+        props_map (mapping): A mapping of STIX object properties to be checked.
+        version (str): Which STIX2 version the properties must confirm to.
+
+    Raises:
+        ValueError: If the properties do not conform.
+    """
+    try:
+        ref_prop_types = _VERSION_REF_TYPES_MAP[version]
+    except KeyError:
+        ref_prop_types = _UNKNOWN_VERSION_REF_TYPES
+
+    for prop_name, prop_obj in props_map.items():
+        tail = prop_name.rsplit("_", 1)[-1]
+        if tail == "ref" and not isinstance(prop_obj, ref_prop_types):
+            raise ValueError(
+                f"{prop_name!r} is named like a reference property but is not "
+                f"a subclass of any of {ref_prop_types!r}."
+            )
+        elif tail == "refs" and not all((
+            isinstance(prop_obj, ListProperty),
+            isinstance(getattr(prop_obj, "contained", None), ref_prop_types),
+        )):
+            raise ValueError(
+                f"{prop_name!r} is named like a reference list property but is not"
+                f"a 'ListProperty' containing a subclass of any of {ref_prop_types!r}."
+            )
+
+
+def _validate_props(props_map, version):
+    """
+    Validate that a map of properties is conformant for this STIX `version`.
+
+    Args:
+        props_map (mapping): A mapping of STIX object properties to be checked.
+        version (str): Which STIX2 version the properties must confirm to.
+
+    Raises:
+        ValueError: If the properties do not conform.
+    """
+    # Confirm conformance with STIX 2.1 requirements for property names
+    #
+    # TODO: Should this be 2.1 or anything unknown assuming 2.1+, like we do
+    # for identifier properties above? `version != "2.0"`?
```

Comment:
> TODO: Should this be 2.1 or anything unknown assuming 2.1+, like we do for identifier properties above? `version != "2.0"`?

Reply:
I made this `!= "2.0"` in the current diff with the expectation that we can amend it later if we need. But LMK if you'd prefer the explicit `== "2.1"`.